R/remotes.R

```diff
@@ -132,8 +133,52 @@ renv_remotes_parse_remote <- function(entry) {
 
 }
 
-renv_remotes_parse_finalize <- function(parsed) {
+renv_remotes_parse_git <- function(entry) {
+  pattern <- paste0(
+    "^",
+    "(?:([^@:]+)::)?",  # optional prefix, providing type
+    "(", # URL start
+      "(?:(https?|git|ssh)://)?",   # protocol
+      "(?:([^@]+)@)?",    # login (probably git)
+      # stack overflow: https://stackoverflow.com/a/26987741/9238801
+      "(",  # host start
+        "(?:(?:(?!-))(?:xn--|_{1,1})?[a-z0-9-]{0,61}[a-z0-9]{1,1}\\.)*",
+        "(?:xn--)?",
+        "(?:[a-z0-9][a-z0-9\\-]{0,60}|[a-z0-9-]{1,30}\\.[a-z]{2,})",
+      ")",  # host end
+      "[/:]([-\\.a-zA-Z]+)",             # a username
+      "(?:/([^@#:]+?))?",                # a repository (allow sub-repositories)
+      "(?:\\.(git))?",                     # optional .git extension
+    ")", # URL end
+    "(?::([^@#:]+))?",                 # optional subdirectory
+    "(?:#([^@#:]+))?",                 # optional hash (e.g. pull request)
+    "(?:@([^@#:]+))?",                 # optional ref (e.g. branch or commit)
+    "$"
+  )
+
+  fields <- c(
+    "entry",
+      "type",
+      "url",
+        "protocol", "login", "host",
+        "user", "repo", "ext",
+      "subdir",
+      "pull",
+      "ref"
+    )
+  parsed <- renv_remotes_parse_impl(entry, pattern, fields, perl = TRUE)
+
+  if (!nzchar(parsed$repo))
+    stopf("'%s' is not a valid remote", entry)
+
+  # If type has not been found & repo looks like a git repo, set it as git
+  if (renv_remotes_is_git(parsed))
```

Comment:
I'm still wondering why this check is needed -- if the above regular expression test succeeds, doesn't that imply that this is a `git` remote? Or are there cases where this regular expression will capture some URL / specification that is not actually to be used by `git`? (If so, an example would be helpful)

Reply:
(FWIW I think the implementation may be good as-is, I just want to double-check)

Reply:
there are cases where the regex would capture non-git URLs - it's able to capture a lot of URLS, for example:

```r
renv:::renv_remotes_parse_git("http://cran.com/non-git")
```

would yield:

```
$entry
[1] "http://cran.com/non-git"

$type
NULL

$url
[1] "http://cran.com/non-git"

$protocol
[1] "http"

$login
NULL

$host
[1] "cran.com"

$user
[1] "non-git"

$repo
NULL

$ext
NULL

$subdir
NULL

$pull
NULL

$ref
NULL
```

Reply:
> > Would the maintainers be okay with hosting test repos, similar to [kevinushey/skeleton](https://github.com/kevinushey/skeleton)?
> FWIW I've now forked those two repositories to my own account, if you want to use those versions in tests.

Great, I'll update the entries in `renv.git1`'s `Remotes:` field to match your urls & put in a PR or however it works when you own the upstream